dit-core/src/peer/proto.rs

```diff
@@ -0,0 +1,182 @@
+
+    /// Append the `SocketAddr` to the DHT entry at `dst`.
+    PutRequest(SocketAddr),
+
+    /// Indicates a successful put of the `SocketAddr` at `src.dht_addr`.
+    PutResponse(SocketAddr),
+
+    /// Requests the DHT entry at `dst`.
+    GetRequest,
+
+    /// Contains the DHT entry at `src.dht_addr`.
+    GetResponse(Vec<SocketAddr>),
+}
+
+impl Payload {
+    pub fn kind(&self) -> PayloadKind {
+        match self {
+            Payload::Ping { .. } => PayloadKind::Ping,
+            Payload::Pong { .. } => PayloadKind::Pong,
+            Payload::NeighborsRequest { .. } => PayloadKind::NeighborsRequest,
+            Payload::NeighborsResponse { .. } => PayloadKind::NeighborsResponse,
+            Payload::PutRequest { .. } => PayloadKind::PutRequest,
+            Payload::PutResponse { .. } => PayloadKind::PutResponse,
+            Payload::GetRequest { .. } => PayloadKind::GetRequest,
+            Payload::GetResponse { .. } => PayloadKind::GetResponse,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum PayloadKind {
+    Ping,
+    Pong,
+    NeighborsRequest,
+    NeighborsResponse,
+    PutRequest,
+    PutResponse,
+    GetRequest,
+    GetResponse,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
+pub struct Neighbors {
+    pub pred: Option<DhtAndSocketAddr>,
+    pub succ: Option<DhtAndSocketAddr>,
+}
+
+pub struct Codec {
+    config: Arc<Config>,
+}
+
+impl Codec {
+    pub fn new(config: Arc<Config>) -> Self {
+        Self { config }
+    }
+}
+
+impl Decoder for Codec {
+    type Item = Packet;
+    type Error = io::Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        if src.len() < 4 {
+            return Ok(None);
+        }
+
+        let mut length_bytes = [0; 4];
+        length_bytes.copy_from_slice(&src[..4]);
+        let length = u32::from_be_bytes(length_bytes).try_into().unwrap();
+
+        if length > self.config.max_packet_length {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "packet too long",
+            ));
+        }
+
+        if src.len() < length {
+            src.reserve(length - src.len());
+            return Ok(None);
+        }
+
+        let packet = serde_json::from_slice(&src[4..])?;
+        src.advance(length);
+        Ok(Some(packet))
+    }
+}
+
+impl Encoder<Packet> for Codec {
+    type Error = io::Error;
+
+    fn encode(&mut self, packet: Packet, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        let encoded_packet = serde_json::to_vec(&packet)?;
+
+        let length = encoded_packet.len() + 4;
+        if length > self.config.max_packet_length {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "packet too long",
+            ));
+        }
+
+        let length_bytes = u32::try_from(length).unwrap().to_be_bytes();
```

Comment:
Couldn't this unwrap cause a panic?

Reply:
Only if `config.max_packet_length > u32::MAX`. So this config should probably be an `u32` instead of `usize`.